pkg/util/tracingservice/client.go

```diff
@@ -0,0 +1,173 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package tracingservice
+
+import (
+	"context"
+	"sort"
+
+	"github.com/cockroachdb/cockroach/pkg/migration/migrationcluster"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/rpc"
+	"github.com/cockroachdb/cockroach/pkg/rpc/nodedialer"
+	"github.com/cockroachdb/cockroach/pkg/util/ctxgroup"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/quotapool"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing/tracingpb"
+	"github.com/cockroachdb/cockroach/pkg/util/tracingservice/tracingservicepb"
+)
+
+// TraceClient provides an interface for accessing the recordings of inflight
+// spans for a given traceID on a nodes' inflight registry.
+type TraceClient interface {
+	GetSpanRecordings(ctx context.Context, traceID uint64) (*tracingservicepb.SpanRecordingResponse, error)
+}
+
+var _ TraceClient = &localTraceClient{}
+
+// localTraceClient executes the local tracing service's code to fetch inflight
+// span recordings.
+type localTraceClient struct {
+	tracer *tracing.Tracer
+}
+
+// newLocalTraceClient returns a local tracing service client.
+func newLocalTraceClient(tracer *tracing.Tracer) TraceClient {
+	return &localTraceClient{tracer: tracer}
+}
+
+// GetSpanRecordings implements the TraceClient interface.
+func (c *localTraceClient) GetSpanRecordings(
+	_ context.Context, traceID uint64,
+) (*tracingservicepb.SpanRecordingResponse, error) {
+	var resp tracingservicepb.SpanRecordingResponse
+	err := c.tracer.VisitSpans(func(span *tracing.Span) error {
+		if span.TraceID() != traceID {
+			return nil
+		}
+		for _, rec := range span.GetRecording() {
+			resp.SpanRecordings = append(resp.SpanRecordings, rec)
+		}
+		return nil
+	})
+	return &resp, err
+}
+
+var _ TraceClient = &remoteTraceClient{}
+
+// remoteClient uses the node dialer and tracing service clients to fetch
+// span recordings from other nodes.
+type remoteTraceClient struct {
+	traceClient tracingservicepb.TraceClient
+}
+
+// newRemoteClient instantiates a remote tracing service client.
+func newRemoteTraceClient(traceClient tracingservicepb.TraceClient) TraceClient {
+	return &remoteTraceClient{traceClient: traceClient}
+}
+
+// GetSpanRecordings implements the TraceClient interface.
+func (c *remoteTraceClient) GetSpanRecordings(
```

Comment:
Why do we even need this remoteTraceClient type? All it's doing is calling into the embedded traceClient.

Reply:
good point, deleted.